Updated the building connectors documentation (#80)
ConduitIO/conduit-site: docs/connectors/building-connectors/developing-source-connectors.mdx

---
title: "Developing a Source Connector"
sidebar_position: 3
---

> A Source is responsible for continuously reading data from a third party system and returning it in form of a [Record](https://pkg.go.dev/github.com/conduitio/conduit-connector-sdk#Record).

Reviewer: ```suggestion
A Source is responsible for continuously reading data from a third party system and returning it in the form of an [SDK Record](https://pkg.go.dev/github.com/conduitio/conduit-connector-sdk#Record).
```

> You need to implement the functions required by Source and provide your own implementations. Information about individual functions are listed below.

Reviewer: ```suggestion
You need to implement the functions required by the Source interface and provide your own implementations. Information about individual functions are listed below.
```

## `source.go`

This file provides the main functionality of your Source Connector.

### Source Connector Lifecycle Functions

> - **`Source`** Struct: Every Source implementation needs to include an [UnimplementedSource](https://pkg.go.dev/github.com/conduitio/conduit-connector-sdk#UnimplementedSource) to satisfy the interface. This allows us to potentially change the interface in the future while remaining backwards compatible with existing Source implementations.  This struct can be modified to add additional fields that can be accessed throughout the lifecycle of the Connector

Reviewer: ```suggestion
- **`Source`** Struct: Every Source implementation needs to include an [UnimplementedSource](https://pkg.go.dev/github.com/conduitio/conduit-connector-sdk#UnimplementedSource) to satisfy the interface. This allows us to potentially change the interface in the future while remaining backward compatible with existing Source implementations. This struct can be modified to add additional fields that can be accessed throughout the lifecycle of the Connector
```

    ```go
    type Source struct {
        sdk.UnimplementedSource

        config           SourceConfig
        lastPositionRead sdk.Position //nolint:unused // this is just an example
        watcher          *fsnotify.Watcher
        recentlyCreated  sync.Map      // To keep track of recently created files
        createCooldown   time.Duration // Cooldown period after a create event
    }
    ```

> - **`NewSource()`**: A constructor function for your Source struct. Note that this is the same function that should be set as the value of Connector.NewSource. The constructor should be used to wrap your Source in the default middleware. You can add additional middleware, but unless you have a very good reason, you should always include the default middleware.

Reviewer: ```suggestion
- **`NewSource()`**: A constructor function for your Source struct. Note that this is the same function that should be set as the value of `Connector.NewSource`. The constructor should be used to wrap your Source in the default middleware. You can add additional middleware, but unless you have a very good reason, you should always include the default middleware.
```

    ```go
    func NewSource() sdk.Source {
        // Create Source and wrap it in the default middleware.
        return sdk.SourceWithMiddleware(&Source{}, sdk.DefaultSourceMiddleware()...)
    }
    ```

> - **`Parameters()`**: A map of named Parameters that describe how to configure the connector. This map is typically generate using `paramgen`

Reviewer: ```suggestion
- **`Parameters()`**: A map of named Parameters that describe how to configure the connector. This map is typically generates using [`paramgen`](https://github.com/ConduitIO/conduit-commons/tree/main/paramgen).
```

    ```go
    func (s *Source) Parameters() map[string]sdk.Parameter {
        return s.config.Parameters()
    }
    ```

- **`Configure()`**: Validates and stores configuration data for the connector. Any complex validation logic should be implemented here.

    ```go
    func (s *Source) Configure(ctx context.Context, cfg map[string]string) error {
    sdk.Logger(ctx).Info().Msg("Configuring Source...")
    err := sdk.Util.ParseConfig(cfg, &s.config)
    if err != nil {
        return fmt.Errorf("invalid config: %w", err)

>     }

Reviewer: ```suggestion
    }
    // add custom validations here
```
